persist bbox ids when serializing NML

jstriebel opened the pull request:
### Description:
Bounding Box ids are optional in the code, but need to be persisted with NMLs so that the bboxes are parsed correctly when being uploaded. This adds default ids when persisting nmls, and removes id from the BoundingBox class.

### Issues:
- fixes #835

### Todos:
 - [x] Updated Changelog
 - [x] Added Tests

jstriebel commented on webknossos/webknossos/geometry/bounding_box.py:
```diff
@@ -31,7 +33,7 @@ class BoundingBox:
-    name: Optional[str] = "Unnamed Bounding Box"
+    name: Optional[str] = _DEFAULT_BBOX_NAME
     is_visible: bool = True
     id: Optional[str] = None
```
Maybe the id should rather be an int?

fm3 replied:
It’s int32 in the proto definition so int sounds right to me

jstriebel commented on webknossos/webknossos/_nml/parameters.py:
```diff
@@ -53,15 +75,23 @@ def _dump_bounding_box(
         if task_bounding_box is not None:
-            self._dump_bounding_box(xf, task_bounding_box, "taskBoundingBox")
+            next_bbox_id = self._get_max_bbox_id() + 1
+            self._dump_bounding_box(
+                xf, task_bounding_box, "taskBoundingBox", next_bbox_id
+            )
```
Do you know if the id is used for the order of the bounding boxes, or rather the order in the xml file? I'm not quite sure why we have them anyways :shrug:

philippotto commented on webknossos/webknossos/_nml/parameters.py:
```diff
@@ -53,15 +75,23 @@ def _dump_bounding_box(
         user_bounding_boxes = getattr(self, "userBoundingBoxes")
 
         if user_bounding_boxes is not None:
+            next_bbox_id = self._get_max_bbox_id() + 2
```
why + 2 and not + 1?

philippotto commented on webknossos/webknossos/_nml/parameters.py:
```diff
@@ -25,17 +26,38 @@ class Parameters(NamedTuple):
     taskBoundingBox: Optional[BoundingBox] = None
     userBoundingBoxes: Optional[List[BoundingBox]] = None
 
+    def _get_max_bbox_id(self) -> int:
```
Could you add a docstring which explains what the maximum bbox id is (e.g, that only integer ids are considered), as the semantics weren't immediately clear to me because the ids can be strings too apparently.

jstriebel commented on webknossos/webknossos/_nml/parameters.py:
```diff
@@ -25,17 +26,38 @@ class Parameters(NamedTuple):
+    def _get_max_bbox_id(self) -> int:
+        bboxes = []
+        if self.taskBoundingBox is not None:
+            bboxes.append(self.taskBoundingBox)
+        if self.userBoundingBoxes is not None:
+            bboxes += self.userBoundingBoxes
```
They are not, `+=` adds another list (the inplace equivalent would be `bboxes.update(self.userBoundingBoxes)`, whereas `append` only adds a single element.

jstriebel commented:
@fm3 @philippotto I removed the `id` parameter for the `BoundingBox` class as discussed, and how name, color and is_visible are handled through methods. Also added a test for the upload-download roundtrip.
